Fix remaining ILLink warnings in System.Reflection

eerhardt opened the pull request:
This removes the remaining ILLink warnings from System.Private.CoreLib that are not in one of the following areas:

1. COM support
2. EventSource / System.Diagnostics.Tracing
3. StartupHookProvider
4. System.Resources

Contributes to #45623

eerhardt committed:
Add RequiresUnreferencedCode to Delegate constructors

Contributes to #45623

eerhardt committed:
Add RequiresUnreferencedCode to some internal Reflection methods to address ILLink warnings.

eerhardt committed:
Address remaining ILLink warnings in Reflection.Emit.

eerhardt committed:
add using

marek-safar commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/ConstructorBuilder.cs:
```diff
@@ -27,7 +27,7 @@ public sealed class ConstructorBuilder : ConstructorInfo
 
         internal ConstructorBuilder(string name, MethodAttributes attributes, CallingConventions callingConvention,
-            Type[]? parameterTypes, ModuleBuilder mod, TypeBuilder type) :
+            Type[]? parameterTypes, ModuleBuilder mod, [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] TypeBuilder type) :
```
Why All? Shouldn't this be actually nothing

MichalStrehovsky replied:
Reflection.Emit is annotated as requiring all members everywhere - it allows us to also have returns that provide all members (e.g. `private TypeBuilder GetTypeBuilder()` is allowed to return `DynamicallyAccessedMemberTypes.All` on line 46 below). This eliminates the need for suppressions. It's pretty easy to ensure all TypeBuilders we have in the system are annotated as `.All` because that's how they're "born".

Ref.Emit cannot really be used safely unless the entire type is available for all cases that matter (base types, interfaces, etc.).

An "unnecessary" annotation on an internal API feels better than a suppression that requires one to think "is this really safe to suppress now?" and "can it become not safe to suppress in the future?".

vitek-karas commented on src/coreclr/System.Private.CoreLib/src/System/Delegate.CoreCLR.cs:
```diff
@@ -56,6 +57,7 @@ protected Delegate(object target, string method)
         // This constructor is called from a class to generate a
         // delegate based upon a static method name and the Type object
         // for the class defining the method.
+        [RequiresUnreferencedCode("The target method might be removed")]
```
This could be better fixed by adding `[DynamicallyAccessedMembers(DynamicallyAccessedMemberType.PublicMethods | DynamicallyAccessedMemberType.NonPublicMethods)]`, right?

eerhardt replied:
True. Fixing.

eerhardt replied:
I will fix it that way. Note that I need to suppress the warning calling `BindToMethodName`, which is annotated as `RequiresUnreferencedCode`.

vitek-karas replied:
Maybe we should fix that as well - @MichalStrehovsky: I don't see anything wrong with changing the annotation BindToMethodName to add [DAM(AllMethods)] on the methodType....

MichalStrehovsky replied:
I don't mind either way - this constructor and the one on MulticastDelegate is pretty much dead code - I'm not aware of any language compiler generating calls to this constructor from delegate type definitions they generate. It's probably only reachable from IL.

vitek-karas commented on src/coreclr/System.Private.CoreLib/src/System/MulticastDelegate.cs:
```diff
@@ -25,13 +25,15 @@ public abstract class MulticastDelegate : Delegate
         // This constructor is called from a class to generate a
         // delegate based upon a static method name and the Type object
         // for the class defining the method.
+        [RequiresUnreferencedCode("The target method might be removed")]
```
Same comment as on `Delegate` should be possible to fix this with `DynamicallyAccessedMembers`.

vitek-karas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:
```diff
@@ -691,6 +691,7 @@ public sealed override Type[] GetForwardedTypes()
             return types.ToArray();
         }
 
+        [RequiresUnreferencedCode("Types might be removed")]
```
I think it's worth adding a comment that the problem is recursivenes. This method returns all nested types recursively, which `DynamicallyAccessedMember` can't express.

vitek-karas commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeModule.cs:
```diff
@@ -360,6 +361,7 @@ public override void GetPEKind(out PortableExecutableKinds peKind, out ImageFile
             return GetMethodInternal(name, bindingAttr, binder, callConvention, types, modifiers);
         }
 
+        [RequiresUnreferencedCode("Methods might be removed")]
```
Worth a comment that this operates on the module type, and thus can't be annotated.

eerhardt committed:
Respond to PR feedback

Annotate Delegate.BindToMethod with DynamicallyAccessedMembers

Fix up requires unreferenced code comments

eerhardt commented:
I believe I have addressed all feedback. Please take another look.

eerhardt committed:
Update API Compat txt for Delegate.CreateDelegate attribute removal

eerhardt commented on src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt:
```diff
@@ -1,4 +1,11 @@
+CannotRemoveAttribute : Attribute 'System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute' exists on 'System.Delegate.CreateDelegate(System.Type, System.Type, System.String, System.Boolean)' in the contract but not the implementation.
+CannotRemoveAttribute : Attribute 'System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute' exists on 'System.Delegate.CreateDelegate(System.Type, System.Type, System.String, System.Boolean, System.Boolean)' in the contract but not the implementation.
 Compat issues with assembly netstandard:
+CannotRemoveAttribute : Attribute 'System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute' exists on 'System.Delegate.CreateDelegate(System.Type, System.Type, System.String)' in the contract but not the implementation.
```
@Anipik @safern @ericstj - the `RequiresUnreferencedCodeAttribute` was introduced in `net5.0`. I'm not sure how it got into mscorlib and netstandard. Did we update those in the 5.0 timeframe?

Also, can you take a look at this change and tell me if it looks correct?

eerhardt committed:
Fix Mono ILLink warnings

eerhardt commented:
The CI is green. I’ll merge this later today unless I hear anymore feedback.
